Light Theme

toger5 opened the pull request:
## Current state
ready to merge.
(when no-one proposes better squashing.)

there are still two issues. which are not big so shouldn't prohibit from merging but ideas are very welcome:
 - not all icons refresh on color change since they cache the texture in the node (for example button icons)
 - highlighting with modulations works really poorly with dark icons. so ideas how to solve that are appreciated.
   - could write another modulation similar function called: color replace... basically a shader which uses the alpha information + a new defined color
   - could save light icons for the ones that can get highlighted. than change to light when it gets modulated... (I hate that solution)


## Description

This implements everything needed for a light theme.
 - there are new icons for dark and light, both hi and low dpi
 - the svgs to png script now colors the pngs and exports them to different folders for dark and light icons.
 - the SCsub file is is updated so that `editor_icons.gen.cpp` uses has all dark and light icons.
 - `editor_icons.gen.cpp` now takes an argument for the icon color which should get registered in the `editor_register_icons(Ref<Theme> p_theme, bool light_icons)` function.
 - when the theme loads it checks for the base_color and if it is light, it loads the correct icons and the correct font color.
 - added **Light** theme as one of the theme presets.
 - I also simplified some colours. By that process  I removed tab highlighting, because it was duplicate logic for the light theme. It was a lot of additional variables + code for relatively little gain.

<img width="1680" alt="screen shot 2017-08-16 at 23 54 24" src="https://user-images.githubusercontent.com/16718859/29386960-430c1f2e-82de-11e7-955d-2c60bd26b4c1.png">

new icon colors:

<img width="1680" alt="screen shot 2017-08-17 at 22 35 14" src="https://user-images.githubusercontent.com/16718859/29486447-1edd83ae-84e6-11e7-9301-3efa1f7d5239.png">


## current version:
<img width="1680" alt="screen shot 2017-08-24 at 21 37 39" src="https://user-images.githubusercontent.com/16718859/29685668-6bd49f56-8916-11e7-9741-480df3cb4d41.png">

## issues / where I could need help
 - the biggest issue so far, is that some icons don't et updated. if icons are part of a button, it just keeps the icon since the button saves the icon inside the object. I need to call set_icon again on EDITOR_SETTINGS_CHANGED. how can I do that in the most elegant way. 
(there is also `_notify_changed("icon")` what does it do? I Couldn't find out how that function works.)

### tracker
 - [x] file icon still white
 - [x] node icons very little contrast
 - [x] replace section in script editor still light. Probably uses hardcoded colors.
 - [x] connect button inside the import dock.
 - [x] font for the header in inspector is always white.
 - [x] update all icons on theme change.

toger5 committed:
White Theme
 - almost all the colors are generated now. They get adapted based on
   the theme color. All the correct icons are used
 - error label now uses error color
 - added missing button colors in editor theme

djrm commented on editor/icons/source/icon_GUI_tree_arrow_right.svg:
```diff
@@ -71,7 +71,7 @@
      id="layer1"
      transform="translate(0,-1040.3622)">
     <path
-       style="fill:none;stroke:#ffffff;stroke-width:2;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:0.39215687"
```
these are not supposed to change, you should create another conversion, like #FFFFFF -> #000000

toger5 replied:
why do they need to be #ffffff and can't be #e0e0e0 if I add #FFFFFF to #000000 than there would be another additional colors changed which might should star the same. because #FFFFFF is used really commenly.

djrm replied:
@toger5 yeah, just preserve the original colors, and create conversions accordingly, ti is white because how it has to look

toger5 replied:
@djrm it actually looks pretty nice with a slightly darker color too. it's only a tiny bit darker. I don't know what issues I will get when adding ffffff -> 000000.
we maybe could do fefefe. so all colors which should become dark use fefefe and all colors which should stay white use ffffff

HummusSamurai commented:
Regarding the section "issues / where I could need help", I think it should be opened in its own issue and this PR merged when the reviewers are done :b 

(aka: It is already very pretty so let us use it!)

hubbyist commented:
This theme needs a little contrast I think. thumbnails in file system dock seems as if merged with the background. Also in the scene panel , icon green and background gray are to similar in brightness so lacking contrast.

HummusSamurai commented:
I agree with @hubbyist that the red and green icons need to be darker, but I am not sure what can be done for the file system.

toger5 commented:
For the filesystem icon i can do the same than for the rest of the icons. Make it dark for the light theme.
I didnt do that yet because it is part of the default theme... (i think, it also could be in the special icons which dont get converted yet. If its in the special icons it would be even easier ! 

Maybe i just add it to the editor theme.

For the node icons we have two options:
Changing overall color ? Maybe also the dark theme benefits from higher saturated theme colors?
Or: it would also be pretty easy to just chnage the color for the bode icons when using the light theme.

I will work on the last icons which are still white + some hardcoded colors. Ill update the first post with a tracker what is still not eorkin in the white them. (help appreciated)

toger5 commented:
I fixed a couple of the tracker issues.

_(@akien-mga I will rebase -i however you desire when I'm completely done. For now I keep it to see the changes in order)_

### These are the node colors used for light themes:
DARK_COLOR_NODE_RED = `#ba5555`
DARK_COLOR_NODE_BLUE = `#576fc1`
DARK_COLOR_NODE_GREEN = `#4eb257`
DARK_COLOR_NORMAL = `#101010`
DARK_COLOR_INTENSE = `#000000`

Feel free to propose whatever you like. I just (more or less random picked)
@djrm you might have an opinion about those colors which is much better than this!

### As a comparison the current colors which dont really work in a light theme:
REPLACE_COLOR_NODE_RED = `#fc9c9c`
REPLACE_COLOR_NODE_BLUE = `#a5b7f3`
REPLACE_COLOR_NODE_GREEN = `#a5efac`
REPLACE_COLOR_NORMAL = `#e0e0e0`
REPLACE_COLOR_INTENSE = `#fefefe`

Geequlim commented:
If you want to make icons darker maybe you need a script to do that like this
https://github.com/Geequlim/godot-themes/blob/master/addons/arc_light/assets/darkenIcons.py

Geequlim commented:
And why not make an external theme ?

djrm commented:
@Geequlim now it is possible to set the editor color to white, making the icons look bad, so we either need to remove the option to choose white backgrounds or ship dark icons and adapt to the white background.

HummusSamurai commented:
@djrm I think both are unnecessary. As long as the default theme options are fine, it should not be the editor's responsibility for the user to choose reasonable theming. This problem you mentioned only arises when the user himself decides to use a white background.

djrm commented:
@HummusSamurai exactly, and since we are giving him the option it means that we have to guarantee that it is a good experience, the other case is what i described as removing the option to choose a white background.

HummusSamurai commented:
@djrm I don't think it's a good idea to remove the option for custom coloring.

I think the most sane way to handle this would be for the user to choose a preset theme, either dark, light, or whatever, which determines the icons that will be applied.
Then he can change the background color (using a color picked), which should change the selected theme to "Custom" instead of the current preset name.

It's simple enough and it doesn't limit the user's choices. I don't believe changing icon themes dynamically is sane behaviour at all, in my humble opinion. It could lead to user inconvenience: at what threshold is "light" considered "light", what about users who want less contrast for some reason?

The only good experience we need to guarantee is that of the default presets; forcing our visual standards even for custom theming is a bit.. extreme.

toger5 commented:
@Geequlim i am using a script.
Each svg gets converted and before the conversions colors get adapted for light and dark colors.
Than we have 4 different png's for light hidpi, light lodpi, darkhidpi, dark lodpi.

@hummusSamurai you bring up good points.
 - a lot of users just use dark. That is true i guess, but on some operating systems, for presentations, and a lot of other edge cases (like weired personal preferences ;) ) light just fits better.
 - forcing users: the point is true. But when bringign the color to lighter than average 0.5 you will have a hard time reading the font itself. So that the font changes black and white dependabd on the background is what users expect. There is not much left to preferences. Either you can read it or you cant. I was just looking at the icons as the same issue. Since most icons are white they also are invisible when using a light color. So i treated them the same than the font.
So i think this wouldnt change that: Light colors have a dark font + dark icons.
BUT the color for the node icons are debateable. They are still hardly visible. So some users might like that.
Forcing a more contrasty icon color maybe is to much.
So we could add another option for `darker-saturated node icons`.
Than users have the option for the them color, automatic font color adaption and users can decide themselves if they want dark node icons.
But honestly i dont see a reson why you want light icons with a light theme color. I tried that and it always looks unusable...

Just as additional information:
This pr is not because i think light theme is necassary. But for a well rounded experience. Noone appreciates the option for a light theme when it is compleatly broken. @djrm explained that pretty well.
And when godot claims to have a easily coustomizable theme it shouldnt look compleatly broken when using a lighter color.
All provided options should be balanced in a way that users have a hard time making it look unusable.

HummusSamurai commented:
@toger5 I agree with everything you said.

neikeq commented:
Just as an addition, when you take your laptop outside in the morning, you don't want to use a dark theme.

djrm commented:
@toger5 also regarding white theme, there are lots of considerations, like having a custom 
* code editor theme
* custom docs colors
* VS nodes colors (maybe)

toger5 commented:
@djrm that is true. I think we have to decide for a general approach:
### option 1:
 - everything is configurable. so the user can also make things that look bad but he at least has all the power he might want
### option 2:
 - the user gets access to some properties. the one we decide are interesting if changeable. but most of the stuff is automatic. so a light theme just works and looks good without too much configuration. 

I honestly prefer option 2... a light theme might just change settings like code editor theme. and when changing back to dark it also changes back...
colors, icons, if highlight color is darker tan theme color or lighter all gets changed automatically.

YeldhamDev commented:
Giving my two cents, I would rather have option 1. I like when software gives me liberty, even if I end up doing stupid things with said liberty, I think it's a fair price.

toger5 commented:
@YeldhamDev would you prefer if you have the option to change everything. but when you change to a light theme godot asks you:
"do you also want to change the icon theme to dark + the code editor to light for a better experience"
or do you just want options and pick the ones that make most sense to you.

YeldhamDev commented:
Automatically, so there is no need for an extra click from the user.

toger5 commented:
@YeldhamDev so basically just exposing the options but changing them with the color change?

YeldhamDev commented:
@toger5 Pretty much. Like I said, those are my two cents, I can see why you and other people may think this is redundant, but I think it's good to have these types of choice available.

hpvb commented:
If I can put my 2 cents in the hat: I personally much prefer light themes for the software I use. If a software has the option to switch away from the 'dark theme' I basically always use it.

I have no thoughts on which is 'best' but I do know I strongly prefer it. I'd be very happy if this gets merged.

djrm commented:
@toger5 if i understood you well, 1 means you can set all the colors (icons color, help colors etc.)?
if thats so option 2 is definitely the way to go

toger5 commented:
@djrm @hpvb @YeldhamDev 
I thought about this a little more, and making everything configurable is not a good choice. Especially in the situation where ppl switch there theme A lot, It would make it too complex to change back and forth between light and dark. It has to be like a preset. I want to be able to change to light and dark with one click without adapting every singe aspect (color, code theme, icon color, font color...)

toger5 commented:
We can add options for the things that you guys really want to be configurable later on. but as a first implementation it is more important that it is easy to get a light theme without knowing each ui setting.

hpvb commented:
@toger5 I'm all for a simple light/dark theme switcher. 👍 💟  etc. 😃

toger5 commented:
there we go!
since svg pr is merged there are thousands of conflicts. so I will change some things to make it work again tomorrow.
light theme is much nicer to implement now anyways!

toger5 committed:
moved 'scale import setting' into svg section

volzhs commented:
this one looks like a light theme of photoshop.
such a great job.
I love @toger5 @djrm

djrm commented:
@toger5 here is the color conversion, table, there are some icon that currently are not converted, like the color ramp, also there is a list of icons that should not be converted.

```
colors = {
	"#e0e0e0": "#4f4f4f", # common icon color
	"#ffffff": "#000000", # white
	"#b4b4b4": "#000000", # script darker color

	"#cea4f1": "#bb6dff", # animation
	"#fc9c9c": "#ff5f5f", # spatial
	"#a5b7f3": "#6d90ff", # 2d
	"#708cea": "#0843ff", # 2d dark
	"#a5efac": "#29d739", # control

	"#ff8484": "#ff3333", # error
	"#84ffb1": "#00db50", # success
	"#ffd684": "#ffad07", # warning

	# rainbow
	"#ff7070": "#ff2929", # red
	"#ffeb70": "#ffe337", # yellow
	"#9dff70": "#74ff34", # green
	"#70ffb9": "#2cff98", # aqua
	"#70deff": "#22ccff", # blue
	"#9f70ff": "#702aff", # purple
	"#ff70ac": "#ff2781", # pink

	# audio gradient
	"#ff8484": "#ff4040", # red
	"#e1dc7a": "#d6cf4b", # yellow
	"#84ffb1": "#00f010", # green

	"#ffd684": "#fea900", # mesh (orange)
	"#40a2ff": "#68b6ff", # shape (blue)

	"#84c2ff": "#5caeff", # selection (blue)

	"#ea686c": "#e3383d", # key xform (red)
}


exceptions = ["icon_editor_pivot", "icon_editor_handle", "icon_editor_3d_handle", 
	"icon_godot", "icon_panorama_sky", "icon_procedural_sky", "icon_editor_control_anchor",
	"icon_default_project_icon"
]
```

toger5 commented:
@djrm this is great!! ty 
I littraly just finished (still does not work entirely though) the auto conversion.
so ill need that list!
this is how It will look in cpp. since it is pretty simple we can use the source code as reference to the color conversion.
```
Dictionary dark_icon_color_dictionary;
//convert color:                              FROM       TO
ADD_CONVERT_COLOR(dark_icon_color_dictionary, "#ffffff", "#000000");
ADD_CONVERT_COLOR(dark_icon_color_dictionary, "#e0e0e0", "#000000");
```

toger5 commented:
I just pushed the current version which uses the svgs implementation.
I can't get It to work though. for some icons the color conversion does not take effect. currently I only replace the #e0e0e0 color.

@djrm maybe you could take a look at it and tell me if you have an idea what is broken.
for debugging perpes most colors replacements are commented out.

toger5 commented:
## update:
most of it works now. it is totally usable. but I think some colors are still not converted.
the biggest issue I was encountering was that the colors from nsvg were saved as RRGGBB
and I expected RRGGBBAA. since I than was reading the not assigened space as the rr channel it all got messed up.

karroffel commented:
If it's ready for merging (aka public testing 😀 ) then change the title back, then we'll see :)

pwab commented:
Thanks **a lot** for this!
Just a quick question slightly related: Is it possible to change the background color of the scene view (2D and 3D) when using the light theme?

Sample (made with completely no photoshop skills):
![Just a sample with light scene background](https://user-images.githubusercontent.com/19271185/29689491-b0419a4c-8923-11e7-99b6-326f90ea9bc9.png)

(I don't mean to use this as a standard background shading of the light theme. I just want to know, if there are settings I can use to achieve this.)

djrm commented on editor/editor_path.cpp:
```diff
@@ -160,7 +160,7 @@ void EditorPath::_notification(int p_what) {
 					set_tooltip(obj->get_class());
 
-					label_font->draw(ci, Point2i(ofs, (size.height - label_font->get_height()) / 2 + label_font->get_ascent()), name, Color(1, 1, 1), left);
+					label_font->draw(ci, Point2i(ofs, (size.height - label_font->get_height()) / 2 + label_font->get_ascent()), name, get_color("font_color", "Lable"), left);
```
are you sure its Lable?

djrm commented on editor/editor_themes.cpp:
```diff
@@ -159,37 +208,29 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
-	const int border_width = CLAMP(border_size, 0, 3) * EDSCALE;
-
-	Color title_color_hl = base_color;
-	if (highlight_tabs)
```
remember that you can bring back highlight tabs

djrm commented:
@toger5 this PR really needs some squashing, a lot of the commits are outdated and irrelevant, also have you measured how much impact does your color conversion function have, since the current technique of duplicating icons its working i would like to suggest to keep that, and later on we can improve the coloring workflow to avoid duplicated icons for dark and light themes, i think that if your method for convert colors really hit the performance its not woth it, since we are increasing the engine and project load time.

toger5 commented:
@djrm i littraly does nothing. I didnt do a statistically relevent test since the co versio. Time on my laptop is about 0.5 to 0.7 without the color conversion.
With it it also always was inthat range. Most of the ti e at the lower end of the range 0.56 is a really commen time i get.
I spend most of the time adding the color conversion. It was by alnost the biggest task for the light theme so i made sure it makes sense.

@karroffel as djrm mentioned there has to be some squashing done so i didnt remove the wip yet.

@pwab there a setting for the default env + yhe 3d clear color. So adapting those would get you the full ligh experience. 
I wony make it change automatically because the editor theme should not affect the project ppl are working on.

toger5 commented:
@karroffel @djrm 
The white theme works now. and could get merged. I still have to unnecessary commits (the last two).
I wanted to ask to completely remove the svg_to_png.py file? so I didn't squash them yet. 
To which commit should I squash the SCsub change?

<img width="1680" alt="screen shot 2017-08-25 at 19 50 51" src="https://user-images.githubusercontent.com/16718859/29727443-2924cbd6-89d4-11e7-91b3-287c21e68b01.png">

it's hard to see because of the contrast but gradients are working too!
<img width="42" alt="screen shot 2017-08-25 at 19 29 29" src="https://user-images.githubusercontent.com/16718859/29727441-29204836-89d4-11e7-85a1-f8339c48469f.png">
<img width="39" alt="screen shot 2017-08-25 at 19 39 39" src="https://user-images.githubusercontent.com/16718859/29727442-2924454e-89d4-11e7-8258-f538390a387a.png">
(color pick it. it is different)

@akien-mga the color issue we talked about on irc still took me a little bit of time.

**If you are interested:**
nsvg saved the color as RGBA BUT they did it the other way around... AABBGGRR so that they could do:
```cpp
rr << 0 //first
gg << 8 //second
bb << 16 //third
aa << 24 //last ... was the logic they followed.
```
so I needed to swap them first...

```c++
return Color::hex(
		(p_hex & 0xFF) << 24 | //red to << 24
		(p_hex & (0xFF << 8)) << 8 | //green to << 16
		(p_hex & (0xFF << 16)) >> 8 | //blue to << 8
		0xFF //alpha to << 0 and set to 255
);
```

toger5 commented:
pls check Current State in the initial post

djrm commented:
oh i've noticed that icon modulation does not work on light themes

novemberist commented:
There are still some issues with the Visual Script Editor, I think:

![vs_light](https://user-images.githubusercontent.com/7927471/29740440-0f3a7ee0-8a57-11e7-8135-94865466df4e.png)

1. The grid is barely visible (it should probably be darker than the background with the light theme)
2. The "+" "-" and "1:1" buttons are not visible at all unless you hover over them

toger5 commented:
@novemberist Thank you! for the feedback!

@akien-mga It still can get merged I guess. since there are definitely multiple issues remaining which have to be discover first. and they are much easier to discover when ppl can test it.

toger5 commented:
<img width="1047" alt="screen shot 2017-08-26 at 14 25 49" src="https://user-images.githubusercontent.com/16718859/29741329-7d5d7090-8a6a-11e7-8539-73562edc31d4.png">

don't even like the dark grid that much
will give it more transparency...

hpvb commented:
How about we switch to the light theme by default at least until a2 or maybe including a2? We can include a note in the announcement that it's temporary.

novemberist commented:
I think light theme as a default is not a good idea until the docs are at least readable:

![light_docs](https://user-images.githubusercontent.com/7927471/29741785-5e7adbfa-8a73-11e7-9305-71a2965d8145.png)

toger5 commented:
Updated the grid.
<img width="1680" alt="screen shot 2017-08-29 at 12 58 34" src="https://user-images.githubusercontent.com/16718859/29817846-eb1054d0-8cb9-11e7-841a-83d989501983.png">

HummusSamurai commented:
@toger Would it be possible to use a darker shade for this theme? It is currently too bright even for a light theme.

Examples[ for blender](https://www.blendswap.com/blends/view/66393) and [for unity](https://i.ytimg.com/vi/9-07tMz22ik/maxresdefault.jpg).

toger5 commented:
@Zylann it is really easy to change the shade color. it works the same than current custom themes.
Font and icons should be selected by the engine based on you shade settings (although it does not work for all icons yet...)

HummusSamurai commented:
@toger5 It's comforting that I can tweak the background color to be custom, but I was just suggesting that the default shade for the light theme be darker a bit.

(I mean just the background and not the fonts/icons)

toger5 commented:
@HummusSamurai chooses settings are really WIP still. I just use something and change it from time to time. but I take you opinion in consideration when touching the default the next time.

HummusSamurai commented:
@toger5 It is actually not bad and can be gotten used to (even if I mostly use dark themes), but I consider this probably the upper limit for how light it can be :b

toger5 committed:
renamed light_color to contrast_color
 - added tab highlighting
 - other changes to the theme to make use of contrast color

toger5 committed:
added icon updating to most ui elements

toger5 commented:
another update.
A lot (almost all) the icons are now reloaded when changing the theme which makes it feel much less broken. actually feels really solid now.
Also fixed some buttons which were missing from editor themes and some other small things.

@akien-mga It would be great if it can get merged as soon as possible so that I don't have to rebase all the time. (the rabase are really hard to do since I remove all the dark svgs)

so as soon as @djrm is okay with getting rid of all the dark svgs and using the c++ color converter of that pr we should merge.

djrm commented:
@toger5 if the color coverter works exactly the same as the dark icons, then i dont have any problem with this.

toger5 commented:
@djrm I honestly don't exactly know how your converter works. but basically I just replace EVERY color for EVERY svgs path in EVERY gradient based on the table you posted recently.
So It should work exactly the same.

Zylann commented:
Just glancing at the screenshots, it's so white... I can barely distinguish the edges of things

toger5 commented:
<img width="1680" alt="screen shot 2017-08-30 at 02 21 01" src="https://user-images.githubusercontent.com/16718859/29849855-e4b2ee98-8d29-11e7-9d94-1ce623586571.png">
I also turned up the contrast a little bit. Otherwise it made me headache ;)

djrm commented on core/color.cpp:
```diff
@@ -47,6 +47,18 @@ uint32_t Color::to_ARGB32() const {
+
+	return c;
+}
+
```
toger this conversion is not necessary and does not do what it says, it is actually doing RGBA, the thing is. nanosvg uses RGBA but stores the color in an unsigned integer, leaving the alpha at the end, and it actualy ignores what is there, the alpha is handled by the opacity attribute  in SVG. so you dont need this. i think you are over complicating the color conversion.

djrm replied:
in any case the function should be called to_RGBA32, and should not ignore alpha

toger5 replied:
I tried with both directions. And made sure I understand how it needs to be stored.
I think I have to do it this way. It does not work otherwise.
I know that alpha is not part of the color but since it is reading from the right I need to add it to the avoid an offset. of course I could just set it to 0xFF or just shift by 16 but it is more versatile like that...

@djrm feel free to send what you think would work. Maybe I have overseen something.

toger5 replied:
This is what NanoSvg does: AABBGGRR
8 bit alpha, 8 bit BlueChannel, 8 bit GreenChannel, 8 bit RedChannel

I hope you are right that the inversion is caused by unsigned/signed int.
Although I think that makes no difference

djrm replied:
@toger5 the function is correct, the name is not, the function is doing RGBA not ABGR

djrm replied:
you are overlooking this, the nanosvg color format is

```
R  G  B  A
FF FF FF FF
```

djrm replied:
maybe you are thinking about bits in reverse direction
a shift moves the bits towards the left so `1 << 2 = 10`.

if you analize your function well, you will see

djrm replied:
if you are still not convinced this is from nanosvg source code
```
{ "red", NSVG_RGB(255, 0, 0) },
{ "green", NSVG_RGB( 0, 128, 0) },
{ "blue", NSVG_RGB( 0, 0, 255) },
```

djrm replied:
@toger5 sorry i misunderstood, i didnt see this
```
#define NSVG_RGB(r, g, b) (((unsigned int)r) | ((unsigned int)g << 8) | ((unsigned int)b << 16))
```

djrm replied:
@toger5 i think you can use the same formula, it is simpler

djrm replied:
like
```
 ((uint32_t)(r * 255)) | ((uint32t)(g * 255) << 8) | ((uint32t)(b * 255) << 16) | ((uint32t)(a * 255) << 24)
```

toger5 replied:
I wrote some unnecessary long text... which is probably obvious for you anyways. 
but it proves the order of the nsvg color.

I also ask if the change form u_int to int could have caused the issue?
```cpp
return Color::hex(
	(p_hex & 0xFF) << 24 | //red to << 24
	(p_hex & (0xFF << 8)) << 8 | //green to << 16
	(p_hex & (0xFF << 16)) >> 8 | //blue to << 8
	0xFF //alpha to << 0 and set to 255
);
```
I think this is the funciton we are talking about.

p_hex is the value I get from nano svg.
I take the last unshifted 8 bits (looking from left to right). By doing the & opertation.
0xFF is 
0000 0000 0000 0000 0000 0000 1111 1111 (i think)
and the color maybe is: 
        1111 1111 0111 0001 1100 0110 0000 1010
so we get 
        0000 0000 0000 0000 0000 0000 0000 1010
than i shift it by 24 to the left (as you said this increases the int value)

this is the result:
        0000 1010 0000 0000 0000 0000 0000 0000
so we moved the red channel from the end to the beginning

But I honestly dont know how signed/unsigned int are stored. if that would mean we start a number in the center:
        0000 0000 0000 0001 0000 0000 0000 0000 = 1
        0000 0000 0000 0000 1000 0000 0000 0000 = -1
than this might be the reason why I had trouble because im assigning an uint to an int which would result in looping the bytes by 16...
If there negative values work by just using one bit at the front as signing, than that wouldnt really change much.

toger5 replied:
this is literally the same:
```cpp
	uint32_t c = (uint8_t)(a * 255);
	c <<= 8;
	c |= (uint8_t)(b * 255);
	c <<= 8;
	c |= (uint8_t)(g * 255);
	c <<= 8;
	c |= (uint8_t)(r * 255);
```
I just used the same style than the other color conversion function for consistency

toger5 replied:
> @toger5 sorry i misunderstood, i didnt see this

no problem at all. It took me sooo long to finally understand why e0e0e0 worked but all the other colors did not.
(for e0e0e0 it does not matter which oder...)

akien-mga replied:
I'm not an expert in bit shifting, but this ABGR really sounds like we're reading it wrong...

karroffel replied:
It depends on the endianess of the system. In little endian systems this code produces `RGBA` because the `A` part gets shifted to the highest position. On big-endian systems this will be `ABGR`. So for little endian (which end consumer computers are) systems this code is wrong.

karroffel replied:
The more system independent solution would be this

```c
union {
    uint8_t color_values[4];
    uint32_t compound_value;
} converter;

converter.color_values[0] = a;
converter.color_values[1] = b;
converter.color_values[2] = g;
converter.color_values[3] = r;

return converter.compound_value;
```

This unsures that the `A` part is always on the lowest address.

karroffel replied:
It seems like nanosvg doesn't really care about memory layout, they do the encoding anyway, so the intial code looks okay to me.

akien-mga replied:
Alright, I played with my calculator a bit and I confirm that the code in this commit is equivalent to nanosvg's `(r << 0) | (g << 8) | (b << 16) | (a << 24)`.

For some turquoise color (R = 0, G = 128, B = 128, A = 255) we'd thus have:
```
(0 << 0) | (128 << 8) | (128 << 16) | (255 << 24)
= 1111 1111 1000 0000 1000 0000 0000 0000
= AA        GG        BB        RR
```
So yeah, if you read the memory byte by byte, it's indeed AGBR, but it's really just reading the int backwards IMO :)

akien-mga replied:
BTW, `Color::to_32()` and `Color::to_ARGB32()` are exactly the same no?

djrm commented on modules/svg/image_loader_svg.cpp:
```diff
@@ -47,7 +46,57 @@ SVGRasterizer::~SVGRasterizer() {
+			0xFF //alpha to << 0 and set to 255
+			);
+}
+inline void change_nsvg_paint_color(NSVGpaint *p_paint, const Color p_old, const Color p_new) {
```
also here you can pass unsigned ints instead of colors, so before calling convert the colors using to_RGBA32(), and thats it, you dont need to convert back and forth.

djrm replied:
you can then get rid of `nsvg_col_to_godot_color`

toger5 replied:
that is a really good hint. I was adding the to_ABGR32 (can us to_RGBA32... because of the swapped color channel) function at the end and didn't consider a simpler implementation based on it.
Just was happy that it finally worked.
before that I was still trying to fighting against the weird order of color channels of nanosvg colors.

djrm replied:
this is still valid

toger5 replied:
yea should definitely be added with your version. it's much nicer!

toger5 replied:
oh but I'm not sure if nsvg guarantees 1111 1111 for the alpha channel. I think that was another issue. it is more or less random...
so for comparing I need to only compare the last 24 bytes.
by shifting to the left it always fills up with 0000 0000 is that correct?
If so I could shift both:
`0101 0101 0010 0000 0011 1111 0110 1010`
`1111 1111 0010 0000 0011 1111 0110 1010`
->
`0010 0000 0011 1111 0110 1010 0000 0000`
`0010 0000 0011 1111 0110 1010 0000 0000`
they are the same

toger5 replied:
just tested.
and comparing without shifting does not work.
I will try with shifting and than research if shifting always adds 0000 0000  and if so it should work.

toger5 replied:
Okay with shifting it works !

djrm replied:
you can also use a bitmask, like ``` uint32_t masked = uintcolor & 0x00FFFFFF ```

djrm commented:
you can also use a bitmask, like
```
uint32_t masked = uintcolor & 0x00FFFFFF
```

On Tue, Aug 29, 2017, 8:31 PM Timo <notifications@github.com> wrote:

> *@toger5* commented on this pull request.
> ------------------------------
>
> In modules/svg/image_loader_svg.cpp
> <https://github.com/godotengine/godot/pull/10382#discussion_r135951764>:
>
> > @@ -47,7 +46,57 @@ SVGRasterizer::~SVGRasterizer() {
>
>  SVGRasterizer ImageLoaderSVG::rasterizer;
>
> -Error ImageLoaderSVG::_create_image(Ref<Image> p_image, const PoolVector<uint8_t> *p_data, float p_scale, bool upsample) {
> +inline Color nsvg_col_to_godot_color(int p_hex) {
> +	//swaps the color from ABGR to RGBA (nsvg is usging ABGR)
> +
> +	return Color::hex(
> +			(p_hex & 0xFF) << 24 | //red to << 24
> +			(p_hex & (0xFF << 8)) << 8 | //green to << 16
> +			(p_hex & (0xFF << 16)) >> 8 | //blue to << 8
> +			0xFF //alpha to << 0 and set to 255
> +			);
> +}
> +inline void change_nsvg_paint_color(NSVGpaint *p_paint, const Color p_old, const Color p_new) {
>
> Okay with shifting it works !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/10382#discussion_r135951764>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDYGVV6eqOdYGwcZ2L082nokQLUilJjks5sdLuIgaJpZM4O5VHF>
> .
>

toger5 commented:
@djrm true would just be longer...
`p->color << 8 == p_old.to_ABGR() << 8`
compared to:
`p->color & 0x00FFFFFF == p_old.to_ABGR() & 0x00FFFFFF`
would id be easier to read?

djrm commented:
@toger5 thats also a reason why im against adding that method to the Color class, that color convension is very weird and i would say is nanosvg specific, you could add such function in the loader class, and then simply 0 the alpha, and then do a regular comparison

toger5 commented:
@djrm with the naming the function in the color class is realtivly straight forward. but I totally see the advantage of doing it inside the loader. (it won't be used for anything else than nanosvg)

But I think the color functions are a little broken anyways
`to_32()` and `to_RBGA32()` does the same thing...
I might remove it in this pr and than propose the other color fixes in another pr.

djrm commented:
@toger5 is a one liner anyway, i would suggest you to add it as an inline function in nanosvg, it would make everything simpler.

akien-mga commented on editor/editor_themes.cpp:
```diff
@@ -86,14 +86,14 @@ static Ref<StyleBoxFlat> change_border_color(Ref<StyleBoxFlat> p_style, Color p_
 	bool is_gizmo = String(editor_icons_names[p_index]).begins_with("Gizmo");
-
-	ImageLoaderSVG::create_image_from_string(img, dark_theme ? editor_icons_sources[p_index] : editor_icons_sources_dark[p_index], EDSCALE, true);
+	
```
Trailing space.

akien-mga commented on modules/svg/image_loader_svg.cpp:
```diff
@@ -31,7 +31,6 @@
 
 #include "os/os.h"
 #include "print_string.h"
-
```
Don't remove this, it's intended that Godot headers (`"name.h"`) are separate from system headers (`<name.h>`).

akien-mga commented on editor/editor_themes.cpp:
```diff
@@ -162,37 +161,29 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 	}
 
+	//Colors
+	bool dark_theme = ((base_color.r + base_color.g + base_color.b) / 3.0) < 0.5;
```
I think this should be made an option, with three values: Auto (same as here and default value), Black, White (or Dark, Light, as you want).

People have different eyes and different tastes, and not everyone might be happy with the hardcoded limit of the auto behaviour.

toger5 replied:
so doing:
```
bool dark_theme = icon_color_setting == DARK || (icon_color_setting == AUTO && ((base_color.r + base_color.g + base_color.b) / 3.0) < 0.5 );
```

djrm replied:
im not sure about this one, i think it has to be automatic because of the contrast also it would add another step to normal people who cant see white on light grey.

akien-mga replied:
That's why I propose to have an Auto mode which is the default value, and behaves as implemented already. But I do think that we'd have users trying a greyish theme and not happy that they can't decide by themselves if they want the dark icon set + font color or the light one (in cases where both might be contrasted enough, or in cases where the auto logic might not pick the best setting).

akien-mga commented on misc/scripts/svgs_2_pngs.py:
```diff
@@ -10,8 +10,31 @@
 import rsvg
 import cairo
 
+
```
You change this script in the first commit, then deprecate it. Wouldn't it be better to just remove it?

akien-mga commented on editor/editor_path.cpp:
```diff
@@ -160,7 +160,7 @@ void EditorPath::_notification(int p_what) {
 					set_tooltip(obj->get_class());
 
-					label_font->draw(ci, Point2i(ofs, (size.height - label_font->get_height()) / 2 + label_font->get_ascent()), name, get_color("font_color", "Lable"), left);
+					label_font->draw(ci, Point2i(ofs, (size.height - label_font->get_height()) / 2 + label_font->get_ascent()), name, get_color("font_color", "Label"), left);
```
That part should ideally have been amended into the commit which introduced the bug.

akien-mga commented on editor/code_editor.cpp:
```diff
@@ -92,6 +92,13 @@ void FindReplaceBar::_notification(int p_what) {
 	} else if (p_what == NOTIFICATION_VISIBILITY_CHANGED) {
 
 		set_process_unhandled_input(is_visible_in_tree());
+	} else if (p_what == EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED) {
```
This seems to be overkill to me. It will reload icons every single time settings are changed, even if those are not icon related settings. You should check what settings changed, and only reload if a relevant themeing one was modified.

djrm replied:
@akien-mga this does not reload the icons, i' ve added a check in editor themes to only reload/regenerate when needed, but yeah it gegenerates the theme, fonts and all styles, there might be a better solution

toger5 replied:
@akien-mga  I needed to add those icons overrides.
Icons are saved in the controls themselves so changing the theme does not change the controls.
So I do a lookup and a set_icon for each control which saves the icon itself and does not load in NOTIFICAITON_DRAW.

But it is correct that there are a LOT of things happening now when changing the theme...
So it would be nice to have a better solution.
We could add another EditorSettings:: ENUM (`EDITOR_THEME_SETTINGS_CHAGNED`)
and than we call sent different notifications.
But since the editor Settings are so generic (just categories via strings that can appear by just one line somewhere...) It is hard to decide when to call which callback without hardcoding the strings that trigger it.

What about adding a static method to EditorSettings : `String EditorSettings::get_chagned_setting()`.
(may that already exists...)
than we could encapsulate as much as possible in the NOTIFICATION_EDITOR_SETTINGS_CHANGED cases so that only the needed things happen... it's some work though...

akien-mga replied:
Yeah, maybe that's outside of the scope of this PR and something to work on after it's merged.

akien-mga commented:
I'm mostly OK to merge, but as always I'd like to have a commit history that tells a readable story. This PR was created as an iterative process, changing implementation details all along (e.g. starting by tweaking the logic for generating dark icons, then marking that same logic as depracated, removing the icons and converting the colors directly in the SVG at runtime). That's good, but for the final result, ideally, it should bring the code base from one state to another state in the most straigthforward way.

So as I read it, this PR should:
- Delete the existing dark icons and remove the script that generates them (as I understand it, they weren't used yet anyway, so it wouldn't break the engine at this stage)
- Implement conversion of SVGs from light variants to dark variants
- Implement the light theme, using those new dark icons
- Some cleanups for readability, that are not part of the new logic but make the code clearer (e.g. c748960b and 2df934b). Those are good as separate commits.

So I'd imagine a good commit history as something like this:
```
- Put scale import setting into svg section (current 2df934b)
- Implement auto color conversion for dark variants of SVG icons (implement the new logic, does not remove the obsoleted icons and script)
- Remove obsoleted dark SVG icons and PNG generation script (i.e., keep all the noise about the thousands of removed lines of SVG code in a separate commit - it's not relevant to read the previous commit)
- Implement light theme using the new dark icons logic (squash of 61a730c + 56ddc95 + 64bd3c0
- Rename light_color to contrast_color (current c748960)
- Reload icons when setting changed (current e8d42d96)
```

Now, I do acknowledge that rebasing to get this kind of clear history is not necessarily trivial at this stage, so I won't really mind if you can't do it. I just wanted to outline the thought process that leads to making good self-documented commits that bring the engine from one coherent state to another coherent state, step by step.

toger5 commented:
@akien-mga I kept the history in mind. As good as I could during working. so it should be possible to achieve the order you propose.

toger5 committed:
removed misc/scripts/svgs_2_pngs.py
 - deprecated because icon conversion now happens in godot itself

toger5 committed:
Implement color conversion for dark SVG icons

toger5 committed:
added to_ABGR32() to core/Color

toger5 committed:
removed svg's in icons/dark

toger5 committed:
added setting for icon/font color: auto,dark,light

toger5 commented:
@akien-mga
 - added the setting:
<img width="970" alt="screen shot 2017-08-30 at 21 32 23" src="https://user-images.githubusercontent.com/16718859/29891264-c34fd012-8dca-11e7-851a-baa981480a45.png">

 - changed history (most importantly separated remove all svgs commit)
 - also improved the icon conversion a little bit. @djrm with your feedback
 - `separated added to_ABGR32() to core/Color` to keep the option to revert it and add an inline function in the svg loader. (if that is preferred)
 - and added some missing icon updates, for script editor (when changing theme)

I also removed the background for the play panel:
<img width="230" alt="screen shot 2017-08-30 at 21 23 12" src="https://user-images.githubusercontent.com/16718859/29891528-afa416f8-8dcb-11e7-9664-0145c2d08c82.png">
I think it looks better and it was a little weird with the white theme. (we can add it back with a StyleBoxFlat with rounded corners.)

The history is currently:
bfcf331 added setting for icon/font color: auto,dark,light
ad11895 added icon updating to most ui elements
a69cfb3 renamed light_color to contrast_color, tab highlighting  - other changes to the theme to make use of contrast color
7818f43 White Theme  - fixed light theme replace dialog  - error label now uses error color  - added missing button colors in editor theme
9151b7c removed svg to pngs
57c6dbf removed svg's in icons/dark
aa8feff Implement color conversion for dark SVG icons
426c6b2 added to_ABGR32() to core/Color
c8448cb moved 'scale import setting' into svg section

For me it displays the `removed svg to pngs` commit at the wrong place on github. (on my branch on github it is correct only in the pr it is broken...) I will try a reword + push -f maybe this fixes it.
